src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLGenerator.java

```diff
@@ -291,22 +318,70 @@ private org.geoserver.mapml.xml.Point buildPoint(org.locationtech.jts.geom.Point
                 .add(this.formatter.format(p.getX()) + " " + this.formatter.format(p.getY()));
         return point;
     }
+
     /**
      * @param p a JTS Polygon
      * @return
      */
     private org.geoserver.mapml.xml.Polygon buildPolygon(org.locationtech.jts.geom.Polygon p) {
+        if (p.getUserData() instanceof TaggedPolygon)
+            return buildTaggedPolygon((TaggedPolygon) p.getUserData());
+
         org.geoserver.mapml.xml.Polygon poly = new org.geoserver.mapml.xml.Polygon();
-        List<JAXBElement<List<String>>> ringList = poly.getThreeOrMoreCoordinatePairs();
-        List<String> coordList =
-                buildCoordinates(p.getExteriorRing().getCoordinateSequence(), null);
-        ringList.add(factory.createPolygonCoordinates(coordList));
+        List<Coordinates> ringList = poly.getThreeOrMoreCoordinatePairs();
+        String coordList = buildCoordinates(p.getExteriorRing().getCoordinateSequence());
+        ringList.add(new Coordinates(coordList));
         for (int i = 0; i < p.getNumInteriorRing(); i++) {
-            coordList = buildCoordinates(p.getInteriorRingN(i).getCoordinateSequence(), null);
-            ringList.add(factory.createPolygonCoordinates(coordList));
+            coordList = buildCoordinates(p.getInteriorRingN(i).getCoordinateSequence());
+            ringList.add(new Coordinates(coordList));
+        }
+        return poly;
+    }
+
+    private org.geoserver.mapml.xml.Polygon buildTaggedPolygon(TaggedPolygon taggedPolygon) {
+        org.geoserver.mapml.xml.Polygon poly = new org.geoserver.mapml.xml.Polygon();
+        List<Coordinates> ringList = poly.getThreeOrMoreCoordinatePairs();
+        ringList.add(new Coordinates(buildTaggedLinestring(taggedPolygon.getBoundary())));
+        for (TaggedPolygon.TaggedLineString hole : taggedPolygon.getHoles()) {
+            ringList.add(new Coordinates(buildTaggedLinestring(hole)));
         }
         return poly;
     }
+
+    private List<Object> buildTaggedLinestring(TaggedPolygon.TaggedLineString ls) {
+        List<Object> result = new ArrayList<>();
+        List<TaggedPolygon.TaggedCoordinateSequence> coordinates = ls.getCoordinates();
+        int last = coordinates.size();
+        for (int i = 0; i < last; i++) {
+            TaggedPolygon.TaggedCoordinateSequence cs = coordinates.get(i);
+            Object value = buildTaggedCoordinateSequence(cs);
+            // client oddity: needs spaces before and after the map-span elements to work
```

Comment:
Now that I think about it, I had noticed that space before/after span tag thing. But it's not working as intended in all cases, it seems.

Reply:
Well duh, this change fixes 3 out of 4 of the cases above, only the countries one is left:

```
            if (value instanceof String) {
                if (i > 0) {
                    value = " " + value;
                } 
                if (i < last - 1) {
                    value = value + " ";
                }
            }
```

Reply:
Fix pushed into this branch.

Reply:
> it's not working as intended in all cases, it seems.

No, there is definitely a bug in the viewer renderer. I am going to fix it, somehow.